### PR TITLE
BMI on Travis for CI

### PR DESCRIPTION
**Pull request by chemistry-sourabh**
Refer #117 

Sets up 3 docker container instead of one as I was having issues with the former. Works perfectly well with all the test cases I have except the picasso test cases which need to be modified to start and stop picasso and einstein as per @naved001's suggestion. Also works with the state operation test cases in #111

### CODE REVIEW

**Comment by ravisantoshgudimetla** [comment]
Thanks @chemistry-sourabh, this is really awesome. Except for end of lines, I don't see much problems with this. But I would recommend this to be tried by @naved001 before we push forward. If not, I can give this a try later this week.

**Comment by chemistry-sourabh** [comment]
Try what ? the travis check passing is proof that it works!

**Comment by ravisantoshgudimetla** [comment]
> the travis check passing is proof that it works!

Where is the travis run?

**Comment by chemistry-sourabh** [comment]
It runs on Travis ? Check the build log for this Travis build, you will understand better

**Comment by naved001** [comment]
@ravisantoshgudimetla  Here are the travis logs. 
https://travis-ci.org/CCI-MOC/ims/builds/232307368#L622

@chemistry-sourabh it's really awesome.

**Comment by chemistry-sourabh** [comment]
Changed to Docker Hub as per long discussion with @ravisantoshgudimetla

**Comment by chemistry-sourabh** [comment]
Fixed new lines also. So please approve if no other issues then we can merge this ASAP

**Comment by chemistry-sourabh** [comment]
Forgot to add coverage and random execution of tests. Will update this PR later

File: .travis.yml [hunk 1]
```diff
@@ -2,10 +2,26 @@ language: python
...
+  - docker
+
 script:
   # Run pep8 on all .py files in all subfolders
   # ignore E402: module level import not at top of file
   - find . -name \*.py -exec pep8 --ignore=E402 {} +
+  - sudo docker run -d -v ceph:/etc/ceph -e MON_IP=172.17.0.2 -e CEPH_PUBLIC_NETWORK=172.17.0.0/24 --name ceph ceph/demo
+  - sleep 30
+  - sudo docker run -d --name hil chemistrysourabh/hil
+  - sudo docker pull chemistrysourabh/bmi-ci
```
Comment (naved001) [inline .travis.yml]: can we have generic names here?

Comment (chemistry-sourabh) [inline .travis.yml]: you mean the repo names ? Ya that is part of my agenda. I didnt knew docker would use my Docker ID as my repo name. Will change it.

**Comment by chemistry-sourabh** [comment]
@naved001 and @sirushtim added randomness plugin, changed repo to bmis instead of mine, Added comments to yaml also. Waiting on approval for coverage

**Comment by chemistry-sourabh** [comment]
@CCI-MOC/bmi-breakers can I get a second +1, so that we can merge. I'll raise a second PR for coverage.

**Comment by naved001** [comment]
Hey @knikolla would you have the time to review this PR?

File: Dockerfile_env [hunk 2]
```diff
@@ -0,0 +1,33 @@
+FROM ubuntu
```
Comment (knikolla) [inline Dockerfile_env]: I don't see this file used anywhere.

File: .travis.yml [hunk 3]
```diff
@@ -2,10 +2,40 @@ language: python
...
+services:
+  - docker
+
 script:
   # Run pep8 on all .py files in all subfolders
   # ignore E402: module level import not at top of file
   - find . -name \*.py -exec pep8 --ignore=E402 {} +
+
+  # Run Ceph Container and Wait for some time so that ceph runs
+  - sudo docker run -d -v ceph:/etc/ceph -e MON_IP=172.17.0.2 -e CEPH_PUBLIC_NETWORK=172.17.0.0/24 --name ceph ceph/demo
```
Comment (knikolla) [inline .travis.yml]: Please use the correct steps as defined here. https://docs.travis-ci.com/user/customizing-the-build#The-Build-Lifecycle

File: Dockerfile_env (hunk 2 above)
Comment (chemistry-sourabh) [inline Dockerfile_env]: This is file is used to build the docker image that is pulled in the yaml file. I'll add a comment on top saying the same.

File: .travis.yml (hunk 3 above)
Comment (chemistry-sourabh) [inline .travis.yml]: I'll have a look

Comment (chemistry-sourabh) [inline .travis.yml]: @knikolla as per the build cycle, the pep8 check should go in the script section and running containers should go in before_install section. But it doesnt make sense to run the containers if pep8 failed right ? what do you think ?

**Comment by ravisantoshgudimetla** [comment]
@chemistry-sourabh please refrain from adding changes once approval has been given unless they are related to change requested or else whoever has approved needs to come back and review it again.

**Comment by chemistry-sourabh** [comment]
@ravisantoshgudimetla sorry about that. I am nearly done with this, will let you know when its ready.

**Comment by chemistry-sourabh** [comment]
@knikolla and @ravisantoshgudimetla please have a look. Added coverage also into this PR.

File: README.md [hunk 4]
```diff
@@ -1,3 +1,5 @@
+[![Build Status](https://travis-ci.org/CCI-MOC/ims.svg?branch=dev)](https://travis-ci.org/CCI-MOC/ims)
+[![Coverage Status](https://coveralls.io/repos/github/CCI-MOC/ims/badge.svg?branch=dev)](https://coveralls.io/github/CCI-MOC/ims?branch=dev)
```
Comment (ravisantoshgudimetla) [inline README.md]: I guess, you need to separate the coverage part out as it is not part of BMI CI. You may be running CI to get coverage report but that wasn't explicitly mentioned in issue. So the steps would be to

- Update the existing issue to reflect this.
- Create 2 commits - One for BMI CI , one for coverage part.

**Comment by ravisantoshgudimetla** [comment]
Overall I am fine with the changes, I will leave it to @knikolla to provide his feedback.

**Comment by chemistry-sourabh** [comment]
@ravisantoshgudimetla are you fine with the changes ? If yes approve else reject

**Comment by ravisantoshgudimetla** [comment]
Refer to my previous comments, @chemistry-sourabh, update the issue and split commits.

**Comment by naved001** [comment]
@knikolla @pgrosu ping

**Comment by chemistry-sourabh** [comment]
@ravisantoshgudimetla updated the issue, but confused on the commit part as some lines are overwritten in the coverage changes so can I just put one commit ? This also doesnt make sense as we will be squashing and merging.

**Comment by ravisantoshgudimetla** [comment]
Single commit may not look good. LMK, where your confusion is. We shouldn't squash if there are 2 commits before merging.

**Comment by chemistry-sourabh** [comment]
@ravisantoshgudimetla and @knikolla done! have a look

**Comment by chemistry-sourabh** [comment]
@ravisantoshgudimetla @knikolla pinging. If you guys are too busy just give a blind approve :P

**Comment by naved001** [comment]
@chemistry-sourabh 

So @pgrosu and I had a look at it. It seems you are starting a container with Ceph and HIL. What version of ceph is that? Because we should document what version of ceph our tests will work with. Thanks!

@pgrosu is there something else that you want to ask before we merge this?

**Comment by chemistry-sourabh** [comment]
@naved001 the ceph container is not made by me. I pull it from the ceph repo, so please check the version at dockerhub.

**Comment by pgrosu** [comment]
So we want the docker container versions of Ceph and HIL to match what we expect to be in our production machines.

**Comment by chemistry-sourabh** [comment]
hmmmm, what are the ceph client and ceph version numbers ?

**Comment by chemistry-sourabh** [comment]
Also what is the hil production version ?

**Comment by chemistry-sourabh** [comment]
Let me know which tag I should use for ceph https://hub.docker.com/r/ceph/demo/tags/

**Comment by pgrosu** [comment]
@chemistry-sourabh Our most up-to-date BMI environment is the VM running at Northeastern, where we perform our development/staging.  This - as I understand - is our closest "production" environment to date.  The Ceph version there can be determined as follows:

```
# ceph-bmi-dev --version
ceph version 0.94.9-9.el7cp (b83334e01379f267fb2f9ce729d74a0a8fa1e92c)
#
```

It is basically the Hammer LTS release from August/September 2016, which can be found here:

http://docs.ceph.com/docs/master/releases/

HIL will be upgraded soon by @naved001, but if you want to know the version it's the following from the Northeastern one:

```
$ cat /usr/lib/python2.7/site-packages/haas-0.2rc2-py2.7.egg-info/PKG-INFO | grep ^Version
Version: 0.2rc2
$
```

HIL is on RHEL 7.3, and BMI is on RHEL 7.2, which would be nice to become consistent.

Thus regarding the tag for Ceph container to use for testing, either one of the following two should be compatible:

* `tag-build-master-hammer-centos-7`

* `tag-build-master-hammer-fedora-23`

Does everyone else agree @naved001, @djfinn14, @apoorvemohan, etc.?

Also @naved001 can you please have the HIL team implement versioning.  It looks like you have releases at the following link - which are versioned:

https://github.com/CCI-MOC/hil/releases

But at the very least please add a `VERSION` file at the top of the HIL repository that conforms to the semantic versioning, or anything else that is preferred.  We need to do the same thing for ourselves.  Now that we have #60 working, we can build HIL containers from just a Dockerfile.  This way we can then have a `requirements.txt` file that conforms to the our supported versioned configurations, which we would use for deployment/testing.  We need to  come closer to continuous integration so that we are not surprised if users experience something unexpected that we could have tested for.

Thanks,
Paul

**Comment by chemistry-sourabh** [comment]
Thanks @pgrosu since we sort of agreed to support rhel only I will need to change the base OS in the container to centos (I guess we can't use RHEL since its enterprise).

Right now the Dockerfile which I wrote for HIL just does a git clone, which I guess needs to be changed to this particular version.

Also the version for ceph-bmi-dev is the ceph client's version correct ?

**Comment by ravisantoshgudimetla** [comment]
@chemistry-sourabh Another thing that you need to be careful about regarding dockerhub container images - not all of them are secure.

**Comment by pgrosu** [comment]
@chemistry-sourabh   Yes it is the same client - you perform the following - and agree that RHEL containers could not be placed on Travis/DockerHub, but could be an option internally via Jenkins once that's up:

```
# cat `which ceph-bmi-dev`
#!/bin/bash
export CEPH_ARGS="--id henn" && ceph $1 $2 $3 $4 $5 $6 $7 $8 $9
```

@ravisantoshgudimetla  We can have the container launch `yum --security upgrade` before proceeding further with the tests.  What do you think?

**Comment by chemistry-sourabh** [comment]
@naved001 @pgrosu changed ceph version and also changed OS to centos.

The issue is with HIL, the current production is so old that the Dockerfile I wrote for it doesn't work. Also the docs for it are also not that detailed. @naved001 can you tell me the commit to which you will be upgrading production HIL to ? I'll change the docker script to install that instead.

**Comment by naved001** [comment]
@chemistry-sourabh The current HIL master is what I'll use to upgrade. So it'll be [this commit](https://github.com/CCI-MOC/hil/commit/477e12097ce7f379056184d43d8f72dfcbe30b5f)

**Comment by naved001** [comment]
@knikolla are your concerns addressed?

**Comment by pgrosu** [comment]
@chemistry-sourabh Thank you the fixes!  I'm assuming the [a177944](https://github.com/CCI-MOC/ims/pull/119/commits/a177944a7e8f380133be64dc2953826e7ae6f128) and [2a588f7](https://github.com/CCI-MOC/ims/pull/119/commits/2a588f72e2799ecb2e5ff7cc0410747b9858b3bc) commits will pass with the [HIL version](https://github.com/CCI-MOC/hil/commit/477e12097ce7f379056184d43d8f72dfcbe30b5f) that @naved001 recommended?  If so, then let's match to what @naved001 is using as his commit for HIL.  Over time I prefer we shift to a continuous integration, where the HIL instance to test against is the latest stable release - which will also be the deployed version in `BMI-Dev` and `BMI-Prod`.

@naved001 This [HIL commit](https://github.com/CCI-MOC/hil/commit/477e12097ce7f379056184d43d8f72dfcbe30b5f) will be the one installed in new `BMI-Dev` environment @ PRB, right?

**Comment by chemistry-sourabh** [comment]
Updated HIL also and it works!!

**Comment by chemistry-sourabh** [comment]
@naved001 @pgrosu Updated TGT also and all tests pass.

**Comment by pgrosu** [comment]
@chemistry-sourabh This is really great news and thank you for making this happen!  I still need to review it a little bit, but had not had the chance today - I promise very soon.

Many thanks!
~p

**Comment by naved001** [comment]
@ravisantoshgudimetla, ping

**Comment by pgrosu** [comment]
@naved001 One thing to give you a heads up on, this PR contains hammer as the Ceph client and @djfinn14 and I are going to be testing a scenario soon, which might require this PR to be changed by using jewel as the Ceph client.  This is not a scenario that would be easy to figure out if one assumes that Travis is working properly, when it might not.

**Comment by chemistry-sourabh** [comment]
@pgrosu can you describe this scenario a bit more ? I tested BMI with Jewel and with worked fine.

**Comment by naved001** [comment]
@pgrosu 

> which might require this PR to be changed by using jewel as the Ceph client.

yeah, in that case we'll raise a PR to modify the changes introduced by this.

**Comment by pgrosu** [comment]
@chemistry-sourabh Jewel would probably work but we need to still test on PRB in the new dev environment, but this PR is using Hammer.  When we `rbd create` using hammer, we noticed that we had to wait 10 seconds before importing the image into BMI, or it would throw an error.  We wanted to look more carefully if we can replicate this scenario multiple times consistently.  We are in the middle of generalizing the acceptance tests, and perform this test.  That is why it is taking us a bit longer, but wanted to ensure we can confirm this scenario.

We did not want this merged yet as a separate on a passed PR could confuse people where the error is coming from, as Travis would also be used to confirm future submitted PRs.  Such layered troubleshooting might or might not pass, causing confusion where the issue is coming from.

Basically we need all merged PRs to ensure they pass all tests, of which ours is the final one.  So it would be best to wait until we have completed each one properly.

**Comment by chemistry-sourabh** [comment]
@pgrosu interesting! The weird question is why isn't it happening in Travis ?

**Comment by pgrosu** [comment]
@chemistry-sourabh Because I don't believe we are not running that control flow scenario of creating an image first, and then immediately importing it.  @djfinn14 started going into the ceph codebase, but before we drill too deep we wanted to confirm by replicating the scenario.  If that is confirmed, then @apoorvemohan minimum requirements list (https://github.com/CCI-MOC/ims/issues/120#issuecomment-306361757) would need to be updated as well.

**Comment by naved001** [comment]
@pgrosu I understand what you are saying. My reason for going ahead and merging this was because this was sitting for too long and would start to rot just like how the install script is right now ( #60 ). We will continue to discover new bugs and patch them as we discover them instead of doing it in open PRs.

And as for the ceph version (hammer vs jewel) we could open up new issue saying that our CI might need to be updated in the future; that way we can keep track of the issues we might have.

If you'd prefer that we unmerge this PR, that's fine by me. You can merge this after you and @djfinn14 figure out the ceph issues.

**Comment by chemistry-sourabh** [comment]
I'll prefer updating it with a future PR. Since we are not experiencing this issue in Travis, this future PR is not time critical.

**Comment by ravisantoshgudimetla** [comment]
I'd recommend cutting a release before making the changes again. It should be something like:

BMI 0.5 Version supports - TGT XX Version, Ceph YY Version etc. If not, it becomes difficult to manage.

**Comment by pgrosu** [comment]
Though I understand, an unknown-known can still exist and we are beginning to shift as a group to a systematic analysis approach when testing or coding BMI, which includes PRs.  That means analyzing all common scenario combinations, and how BMI handles them.  Once the product is released we want to ensure surprises are minimized downstream, as we will need to focus on the redesign.

**Comment by naved001** [comment]
@pgrosu so in conclusion, what do we want to do with this PR?

**Comment by pgrosu** [comment]
@naved001 In conclusion we will run the acceptance test and report back the status, in case we require the change to jewel.  So let's keep it as is but in the future, please ask before merging.

**Comment by naved001** [comment]
@pgrosu 

> So let's keep it as is but in the future, please ask before merging.

Sure. The PR had 2 +1s which is why I merged it. In future, for any other PR that you don't want merged,  just hit the Request Changes button with a comment.

**Comment by chemistry-sourabh** [comment]
Better would be to add the DO NOT MERGE label
